Introduce Speech portal.

eeejay opened the pull request:
I think there is some confusion about what the speech provider API is and how it relates to libspiel. There are two parts:

1. A speech provider D-Bus interface that any session service can implement. If the session service's name ends with `.Speech.Provider`, apps can discover it, query its voices and synthesize text into audio. The interface defines two properties and one method: Name, Voices and Synthesize respectively. The spec for this interface is housed in its [own repo](https://github.com/project-spiel/libspeechprovider) that also describes a special binary stream format.
2. A client library called `libspiel` that makes the use of a distributed set of providers easy and straightforward for app developers. The API of libspiel is a lot more "evolved" and does not resemble the D-Bus API at all. Some of the features of the library include:
    1. A `Speaker` class that manages the play state of a queue of `Uterance` instances and emits signals for different speech and media events.
    2. An optional GStreamer source to use instead of the default output sink.
    3. Automatic selection of voices to use for each utterance depending on the language of the utterance and the user's preferences.
    4. A live list of available voices across all available providers
    5. A live list of providers that is kept updated by listening to the session bus for `ActivatableServicesChanged` and `NameOwnerChanged` signals.
    6. Other cool things I can't think of at the moment.

The portal API proposed here is aligned with the D-Bus interface above. It does not "proxy" or conform at all to `libspiel`! Because a sandboxed app cannot use or query the session bus the portal API also needs a way for apps to discover speech providers. So the only addition to the portal API over the host's speech provider interface is the `GetProviders` method. Again, this is not a "libspiel" concept, this functionality is a substitute for the session bus access that a host app would have.

## So why does the portal implementation depend on libspiel?
One of the hardest and most complex parts of libspiel is its ability to track available speech providers. It uses the two dbus signals mentioned above and works to keep a live list updated working through different kind of cases (installing uninstalling, activating, killing, and combinations of all). This is a brittle job and prone to bugs. I worked hard to have this be as robust as possible in libspiel and have a pretty comprehensive test suite for this.

Instead of copying all that logic into xdp I chose to rely on the implementation in libspiel and not duplicate the bugs and complexity into another project. Aside from the provider detection code, no other parts of libspiel are used in this portal. The synthesize method is called directly via a dbus proxy and the voices are taken from the proxy as well.

## Why expose providers at all? Why not just show a flat list of voices?
1. The current client implementation already exposes providers to apps. There was a request for that, and we added it to libspiel.
2. I don't think it is the portal's job to further abstract the host APIs and add more logic and complexity for voice management.
3. Exposing the providers allows sandboxed and non-sandboxed clients to share the same model of providers and voices. Why not just use the portal API in non-sandboxed environments? You could. But then you make the portal a hard requirement in the speech stack and I don't think that is wise.
4. My plan is to write a drop-in implementation in libspiel for sandboxed apps, so having a 1:1 relationship for voices and providers makes that easy. Otherwise we will need to deobfuscate and demarshall each voice and its associated provider adding more complexity.
5. We have been working through the speech API and figuring out real world cases for a while, I don't want to re-litigate those decisions and reinvent this all for portals.

## What data is streamed over the file descriptor?
Each voice provides a string that defines its audio format. In its simplest form it just writes raw audio to the file descriptor. The format string is parsed by gstreamer and looks like this:
`audio/x-raw,format=S16LE,channels=1,rate=22050`

Synthesized text can also provide embedded boundary events for when words or sentences or certain text offsets are spoken. For this we have another kind of mime type. A format string for that would look like this:
`audio/x-spiel,format=S16LE,channels=1,rate=22050`

This format is basically a stream of either audio data or event data chunks. The mimetype has the unfortunate name of `audio/x-spiel`, but is not related to libspiel, and it is implemented in a [standalone library](https://github.com/project-spiel/libspeechprovider) so that speech providers don't depend on libspiel.

eeejay committed:
speech: Introduce Speech portal

For now, this consists of speech synthesis, although in the future this
can be expanded to recognition as well.

Mikenux commented:
As for libspiel, I believe that if it is adopted by all DEs, it's acceptable. However, if there are different positions on which library to use, this might be problematic, as it would require supporting multiple libraries (especially for parts used without a corresponding method or option/property in the portal). But this will depend on the maintainers.

There may be a problem. If I understand correctly, you seem to be saying that apps can be providers, not just the system. If so, this is, in my opinion, incorrect because it implies that data (in text form) from one app can be shared with another app. App data can only be shared with the system, not with other apps, and only the system can be a provider. (sorry, I had missed this part)

eeejay commented:
## Shouldn't this be called the libspiel portal?

The speech providers that the client apps are accessing are not based on libspiel but on a simple [standard-ish d-bus protocol](https://project-spiel.org/dbus_iface.html). There is an option for providers to stream a special kind of audio format described above, but doing so is not implemented in libspiel, but a [separate library](https://project-spiel.org/libspeechprovider/) that is designed to be as minimal and portable as possible.

## Yeah, but isn't this still a "spiel" thing, not a generic speech service?

Spiel is the name I chose for the project of a standardized desktop speech platform. It would be the same as referring to the Location portal as the Geoclue2 portal.

## But aren't there other ways for applications to speak with other frameworks?

Yes. There is speech dispatcher, for example. Speech dispatcher is not a general use speech framework but designed for screen readers and other audio-only interfaces. "Spiel" has been adopted by Orca, WebKitGTK, and hopefully Firefox soon. A quick search on flathub reveals that the amount of talking applications is still pretty limited, and there is still a lot of potential for more adoption.

bilelmoussaoui commented:
Making a portal optional is imho a good way for not having any one implement it :)

bilelmoussaoui commented on data/org.freedesktop.portal.SpeechSynthesis.xml:
```diff
@@ -0,0 +1,169 @@
+            * A human readable name
+            * A unique identifier
+            * Synthesis output format
+            * A voice features bit field
```
any chance the bit field could be expanded with it possible values?

eeejay replied:
Yeah, I can add that. Will do that on monday.

bilelmoussaoui commented on data/org.freedesktop.portal.SpeechSynthesis.xml:
```diff
@@ -0,0 +1,169 @@
+        @voice_id: The voice identifier for the voice that should be spoken.
+        @pitch: The voice pitch in which the text should be spoken.
+        @rate: The rate in which the text should be spoken.
+        @is_ssml: True if the text should be interpretted as an SSML snippet.
```
what is an ssml snippet?

eeejay replied:
https://www.w3.org/TR/speech-synthesis/

bilelmoussaoui replied:
would be nice to document it there or at least link to this.

bilelmoussaoui reviewed (commented):
didn't review it fully, just looked through the dbus iface

eeejay commented:
> Making a portal optional is imho a good way for not having any one implement it :)

What is this in response to?

smcv commented:
> Making a portal optional is imho a good way for not having any one implement it

All portals are optional: we can't force desktop environment developers to implement something that they don't want to implement.

bilelmoussaoui commented:
> > Making a portal optional is imho a good way for not having any one implement it
> 
> All portals are optional: we can't force desktop environment developers to implement something that they don't want to implement.

Sorry, should have been clearer. What I meant is having it optional at build time on xdp side.

eeejay commented:
> > > Making a portal optional is imho a good way for not having any one implement it
> > 
> > 
> > All portals are optional: we can't force desktop environment developers to implement something that they don't want to implement.
> 
> Sorry, should have been clearer. What I meant is having it optional at build time on xdp side.

libspiel is a relatively new library and I wanted to give the option to build xdp without it. Should I change that and make it a hard dependency?

eeejay commented:
I took inspiration from geoclue that has a similar build time option.

swick commented:
My 2 cents: I want a speech synthesis portal in xdp. I don't like that libspiel is supposed to be used in xdp when all it has to do is figure out which dbus service to call. This seems like side-stepping what xdp is supposed to do.

DemiMarie commented:
From a security perspective, I’m not a fan of having the string be passed directly to GStreamer without validation.

eeejay commented:
> From a security perspective, I’m not a fan of having the string be passed directly to GStreamer without validation.

You mean the audio format string voices provide? Yeah, we can validate it in the portal and limit what is allowed (for example only audio mime types).

swick committed:
permission: Add lookup and store variants which return futures

When a dbus method call is dispatched in a fiber, those functions allow
us to query or store permissions without blocking and without callback
hell.

swick committed:
request-dex: Add a new Request implementation which uses futures

This also allows us to change the design a bit: requests get created by
the dbus method handler functions and are owned. When the ref-count
drops to zero, the Request gets closed and unexported. This allows using
an autoptr in the handler for ownership, without ever having to owne it
somewhere else.

swick committed:
session-dex: Add a new Session implementation which uses futures

Similar to the Request, this session implementation uses futures and
doesn't need any external ownership. Unlike a Request, a Session is
supposed to be used in multiple calls, so the portal needs to store it
somewhere to look it up from different method handlers. The new
XdpSessionDexStore can be used for this purpose.

swick commented:
Sorry for taking so long to get back to this. A few quick remarks, not a full review:

* I would like to avoid adding new portals which use the old threaded and async model. If you rebase on https://github.com/flatpak/xdg-desktop-portal/pull/1855 you can use libdex instead and make both of our lives easier.
* The libspeechprovider dependency and subproject seems to be left over from when it was required
* We use the Xdp prefix for portals nowadays and the G_DECLARE/G_DEFINE type macros

swick reviewed (commented):
Didn't look at the code yet. The interface seems reasonable overall.

jadahl commented on data/org.freedesktop.portal.Speech.xml:
```diff
@@ -0,0 +1,176 @@
+        @parent_window: Identifier for the application window, see :doc:`window-identifiers`
+        @handle: Object path for the :ref:`org.freedesktop.portal.Request` object representing this call
+        @pipe_fd: File descriptor of pipe to write to.
+        @text: The text to be spoken.
```
I would assume that as of now, speech providers can only be assumed to always part of the system. If/when we would allow installing speech providers as flatpaks, we'd have to have something equivalent to an a11y portal, where providers cant just add themselves.

eeejay replied:
That's not to say that a future provider would not rely on a cloud service for speech generation. We should provide add a flag for that to let the user know it is not a local synth. [Similar to the Web Speech API localService](https://developer.mozilla.org/en-US/docs/Web/API/SpeechSynthesisVoice/localService).

eeejay replied:
I would also consider not allowing "cloud providers" at all, and requiring providers to be completely sandboxed. This does touch a bit on the voice distribution question I brought up a few times. It would disallow providers to provide their own UI/store where users can download additional voices.

eeejay replied:
what if non-system providers (ie. flatpaks) were fully restricted and disallowed network access. Would that address the privacy concern?

swick committed:
session: Take a XdpAppInfo

It already provides the sender and app id. It is also useful because
this way we don't need a lookup from sender to XdpAppInfo anymore.

swick committed:
app-info: Add a destroyed signal

The XdpAppInfo represents a single sender and (ignoring temporary
references) gets destroyed when the sender drops off the bus. That's a
more targeted signal than subscribing to peer_died, and also again means
we don't have to have a lookup from the sender name to the XdpAppInfo.

swick committed:
context: Make dispatching DBus interface skeletons in threads explicit

We're going to add dispatching in fibers as well, and we want to replace
the threads with it, so a default to threads is not desired anymore.

This also changes the XDP_CONTEXT_EXPORT_FLAGS_HOST_PORTAL flag which so
far meant "run on the main thread and don't authenticate". Now that the
default flags mean "run on the main thread" we can drop the host portal
flag and instead add XDP_CONTEXT_EXPORT_FLAGS_SKIP_AUTH to skip
authentication.

swick committed:
context: Make it possible to dispatch DBus interface skeletons in fibers

Fibers allow us to handle method calls in a single function which can
yield by awaiting futures. This means we can get rid of threading
without blocking the main thread.

swick committed:
app-info-registry: Make xdp_app_info_registry_lookup_sender private

The notification portal is changed to use the XdpAppInfo destroyed
signal which gives direct access to the XdpAppInfo and means it doesn't
need the lookup from sender.

The usb portal uses the XdpAppInfo from the session object.

swick committed:
app-info-registry: Use per-sender libdex channels to enforce order

By creating a queue and a worker fiber which processes enqueued items in
order, we get full order which extends to worker threads.

The channel gets destroyed as soon as there is no more pending work and
re-created on demand.

The insert operation takes a DexFuture which resolves to an XdpAppInfo,
which allows to "register" the insert operation and complete it later
while everything stays in order.

swick committed:
dex: Add dex_scheduler_spawnv

Same as dex_scheduler_spawn() but trampolines into a fiber without
having to create special structures on the way there.

This function is already upstream and will be in version 1.2, solets
guard it properly from the buildsystem, so when it becomes available,
everything still works.

swick committed:
ci: Build libdex from commit from main for tests to pass

libdex has a code path which works without io-uring, but that one
creates a thread in the shared object constructor which then triggers:

    g_log_writer_default_set_use_stderr: assertion 'g_thread_n_created () == 0' failed

It also uses g_debug() in the shared object constructor which means it
will get printed (if G_MESSAGES_DEBUG is set appropriately) before the
glib test framework has a chance to redirect it (to add the `# `
prefix). This breaks the test TAP output.

In a later version of libdex, this shared object constructor will get
dropped and both problems will go away.

Ideally there would be a libdex release with the fixes included so we
could just depend on it from meson, but that's not the case. So for now
this just builds from main without bumping the build requirements
because it only affects test and not production.

swick committed:
permission: Get a deep copy from GVariant directly

Instead of getting a shallow copy and then creating a deep copy from it,
we can just get a deep copy via `^as` from the GVariant, which allows us
to steal the value directly.

swick committed:
context: Unexport portals and drain the main loop on dispose

This is done so that fibers get a chance to handle dex_await returning a
cancelled error and shutting things down appropriately.

swick committed:
tests: Add a test checking shutdown with in-flight requests

swick committed:
tests: Make it possible for tests to capture stderr/stdout of xdp

By default this is turned off, so that it lands in the tests logs, but
in some situations we actually want to know about messages being
produced.

eeejay commented on data/org.freedesktop.portal.Speech.xml:
```diff
@@ -0,0 +1,176 @@
+      <arg type="s" name="parent_window" direction="in"/>
+      <annotation name="org.qtproject.QtDBus.QtTypeName.In2" value="QVariantMap"/>
+      <arg type="a{sv}" name="options" direction="in"/>
+      <arg type="o" name="handle" direction="out"/>
```
I decided to have this because maybe we do want an interactive access dialog in the future? This would allow us to do that. I don't know on what basis that would be decided though

eeejay commented on data/org.freedesktop.portal.Speech.xml:
```diff
@@ -0,0 +1,176 @@
+            * A unique identifier
+            * Synthesis output format
+            * A voice features bit field.
+              See here for full list: https://github.com/project-spiel/libspeechprovider/blob/main/interface/org.freedesktop.Speech.Provider.xml#L65
```
I'll change it to a link to actual API docs instead of an arbitrary line number in a github repo.

eeejay commented on data/org.freedesktop.portal.Speech.xml:
```diff
@@ -0,0 +1,176 @@
+      <arg direction="in"  type="s" name="language" />
+      <annotation name="org.qtproject.QtDBus.QtTypeName.In10" value="QVariantMap"/>
+      <arg type="a{sv}" name="options" direction="in"/>
+      <arg direction="out" type="o" name="handle" />
```
Yeah, I think that is the question.. Should we add an access dialog?

eeejay replied:
done. I added access dialogs.

eeejay commented on data/org.freedesktop.portal.Speech.xml:
```diff
@@ -0,0 +1,176 @@
+    <method name="GetVoices">
+      <arg type="o" name="session_handle" direction="in"/>
+      <arg type="s" name="parent_window" direction="in"/>
+      <arg type="s" name="provider_id" direction="in"/>
```
Yeah, I guess so. I could change that.

eeejay commented on data/org.freedesktop.portal.Speech.xml:
```diff
@@ -0,0 +1,176 @@
+      <arg type="o" name="session_handle" direction="in"/>
+    </signal>
+
+    <signal name="VoicesChanged">
```
Sorry, I did `push -f` before I noticed you commented inline here and now I am not sure what you are referring to.

eeejay commented:
Thanks for your patience. This current patch has taken a lot of the feedback given. I might have missed a few points with the chaos of pushing a change before reading the inline stuff. Please let me know if there was anything I missed.

Here are the high level changes:

1. Ported to libdex using @swick's branch. This makes things so much easier! I needed to add a direct constructor to `XdpRequest` since `xdp_request_init_invocation` isn't called on fiber method handlers. The use of `XdpRequestFuture` is not possible since this portal does not have an implementation. Same with `XdpSessionFuture`.
2. Removed the providers getter and signal and stuck to a voices getter and signal. This simplifies the D-Bus API and allows getting everything in one go. I added a `GetProviderDetails` so that additional info on the providers could be retrieved, but it is not necessary for basic functionality.
3. Changed the provided voices to a vardict instead of a struct for extensibility.
4. I added an access dialog.
5. I improved the documentation, and documented the [speech stream format](https://project-spiel.org/libspeechprovider/overview.html).
6. Last, but not least, I introduced an "allow list" in the permissions store. The list details the speech providers that the host deems as trusted and can be exposed via the portal. The mechanics for altering this list are not in this patch, and may require implementation interfaces as mentioned in the comments, but this is out of scope for this PR.

swick committed:
secret: Init async in fiber

We now have libdex which means we can do async initialization which
looks like sync, if we run in a fiber. The secret portal already needs
to run async, so let's port it over to a fiber for init.

swick commented:
Still busy with other things unfortunately, but good to see progress here.

You might want to take a look at XdpRequestDex and XdpSessionDex which should help simplifying things even more. The other thing that I'm a bit unsure about is the enumeration via dbus service names. I really feel like an [interface](https://specifications.freedesktop.org/desktop-entry/latest-single/#interfaces) is the better approach because we can store arbitrary data without having to activate the service.

Either way, I just merged the dex support and I'll also work on making it possible to have experimental portals, so I hope we can get this merged soon.

eeejay commented:
> You might want to take a look at XdpRequestDex and XdpSessionDex which should help simplifying things even more.

I wasn't sure how they simplify this portal. Since the portal has no implementation interface it seemed like more work using those because there would need to be some modification to their constructors. Happy to switch over to them if I am wrong though.

> The other thing that I'm a bit unsure about is the enumeration via dbus service names. I really feel like an [interface](https://specifications.freedesktop.org/desktop-entry/latest-single/#interfaces) is the better approach because we can store arbitrary data without having to activate the service.

Is `Implements` supported in `.service` files? Not all speech providers will have `.desktop` files because they may not have an app/GUI. Would this still be worthwhile?

> 
> Either way, I just merged the dex support and I'll also work on making it possible to have experimental portals, so I hope we can get this merged soon.

Thanks!

swick commented:
> I wasn't sure how they simplify this portal. Since the portal has no implementation interface it seemed like more work using those because there would need to be some modification to their constructors. Happy to switch over to them if I am wrong though.

There are a few good reasons to use them:

* They work without threads. You currently use SESSION_AUTOLOCK_UNREF for example because the old session class requires locking because it internally uses threads.
* No subclassing. You can store the session in your own struct/class however you want.
* Eliminates state tracking. The new session class automatically emits close, calls close for the backend, etc when the ref count reaches zero.

> Is Implements supported in .service files? Not all speech providers will have .desktop files because they may not have an app/GUI.

No, but you can still ship desktop files for non-UI applications with `NoDisplay=true`.

eeejay commented:
Ok, updated request/session to dex versions. Regarding `Implements`, seems like a non-breaking addition we can do in the future. I'll try to tackle that when I have free cycles.

philn commented:
Thank you @eeejay for working on this! 

We have a libspiel backend in WebKit but it's currently disabled because it runs in the sandboxed WebProcess. Having a portal would definitely be welcome :)
